buidl/mnemonic.py

```diff
@@ -26,2090 +30,86 @@ def secure_mnemonic(num_bits=256, extra_entropy=0):
     preseed = randbits(num_bits) ^ extra_entropy
     # convert the number to big-endian
     s = int_to_big_endian(preseed, num_bits // 8)
+    # convert to mnemonic
+    mnemonic = bytes_to_mnemonic(s, num_bits)
+    # sanity check
+    if mnemonic_to_bytes(mnemonic) != s:
+        raise RuntimeError("Generated mnemonic does not correspond to random bits")
+    return mnemonic
+
+
+def mnemonic_to_bytes(mnemonic):
+    """returns a byte representation of the mnemonic"""
+    all_bits = 0
+    words = mnemonic.split()
+    num_words = len(words)
+    for word in words:
+        all_bits <<= 11
+        all_bits += BIP39.index(word)
+    num_checksum_bits = num_words // 3
+    checksum = all_bits & ((1 << num_checksum_bits) - 1)
+    all_bits >>= num_checksum_bits
+    num_bytes = (num_words * 11 - num_checksum_bits) // 8
+    s = int_to_big_endian(all_bits, num_bytes)
+    computed_checksum = sha256(s)[0] >> (8 - num_checksum_bits)
+    if checksum != computed_checksum:
+        raise ValueError("Checksum is wrong")
+    return s
+
+
+def bytes_to_mnemonic(b, num_bits):
+    """returns a mnemonic given a byte representation"""
+    preseed = big_endian_to_int(b)
     # 1 extra bit for checksum is needed per 32 bits
-    checksum_bits_needed = num_bits // 32
+    num_checksum_bits = num_bits // 32
     # the checksum is the sha256's first n bits. At most this is 8
-    checksum = sha256(s)[0] >> (8 - checksum_bits_needed)
+    checksum = sha256(b)[0] >> (8 - num_checksum_bits)
     # we concatenate the checksum to the preseed
-    total = (preseed << checksum_bits_needed) | checksum
+    all_bits = (preseed << num_checksum_bits) | checksum
     # now we get the mnemonic passphrase
     mnemonic = []
     # now group into groups of 11 bits
-    for _ in range((num_bits + checksum_bits_needed) // 11):
+    for _ in range((num_bits + num_checksum_bits) // 11):
         # grab the last 11 bits
-        current = total & ((1 << 11) - 1)
+        current = all_bits & ((1 << 11) - 1)
         # insert the correct word at the front
-        mnemonic.insert(0, WORD_LIST[current])
+        mnemonic.insert(0, BIP39.word(current))
         # shift by 11 bits so we can move to the next set
-        total >>= 11
+        all_bits >>= 11
     # return the mnemonic phrase by putting spaces between
     return " ".join(mnemonic)
 
 
-WORD_LOOKUP = {}
+class WordList:
+    @classmethod
+    def _load(cls):
+        word_file = path.join(path.dirname(__file__), cls.filename)
+        with open(word_file, "r") as f:
+            cls.words = f.read().split()
+        cls.lookup = {}
+        for i, word in enumerate(cls.words):
+            # add the word's index in the dict lookup
+            cls.lookup[word] = i
+            # if the word is more than 4 characters, also keep
+            #  a lookup of just the first 4 characters
+            if len(word) > 4:
+                cls.lookup[word[:4]] = i
+
+    @classmethod
+    def index(cls, word):
```

Comment:
Maybe `get_index()` would be less ambiguous? On second thought, `get_index()` could imply that if there wasn't a matching word you return `None` instead of raising an exception.

I think the correct way to call this might just be `WordList.lookup[word]` (get rid of the helper method all together) instead of `WordList.index(word)`. This is why I find classes tricky, because the caller has to look deep under the hood to figure out how they work. The previous `BIP39_WORD_LIST[word]` seems strictly superior to the new class-based structure.

Reply:
Also, should this lookup be for `word.lower()` to be more forgiving?

I was thinking that would happen as part of the `normalize()` function, but in the current code-path this method is called first from `from_mnemonic()`.

Reply:
I was thinking `.index` is a method available to lists. Maybe I should subclass a data structure. Let me see if I can make it more usable that way.